### PR TITLE
[torchlib] Implement missing operators (set1)

### PR DESCRIPTION
**Pull request by shubhambhokare1**
Implement missing operators uncovered by torch.onnx tests as per #1644

- [x] Implement <OpOverload(op='aten.fmod', overload='Scalar')>
- [x] Implement <OpOverload(op='aten.fmod', overload='Tensor')>
- [x] Implement <OpOverload(op='aten.glu', overload='default')> @shubhambhokare1 
- [x] Implement <OpOverload(op='aten.le', overload='Scalar')>
- [x] Implement <OpOverload(op='aten.lerp', overload='Scalar')>
- [x] Implement <OpOverload(op='aten.linalg_cross', overload='default')>
- [x] Implement <OpOverload(op='aten.mv', overload='default')>
- [x] Implement <OpOverload(op='aten.pow', overload='Scalar')>

- [x] Implement <OpOverload(op='aten.remainder', overload='Scalar')>
- [x] Implement <OpOverload(op='aten.remainder', overload='Tensor')>
- [x] Implement <OpOverload(op='aten.silu', overload='default')>
- [x] Implement <OpOverload(op='aten.unsafe_split', overload='Tensor')>

[**NOT PART OF THIS PR**] Requires adding implementation functions in torchlib eventually (not currently high in priority)

- [ ] Implement `<OpOverload(op='aten.__rshift__', overload='Scalar')>`
- [ ] Implement <OpOverload(op='aten._linalg_det', overload='default')>
- [ ] Implement <OpOverload(op='aten._linalg_slogdet', overload='default')>
- [ ] Implement <OpOverload(op='aten._prelu_kernel', overload='default')>
- [ ] Implement <OpOverload(op='aten.add', overload='Scalar')>
- [ ] Implement <OpOverload(op='aten.add', overload='Tensor')>
- [ ] Implement <OpOverload(op='aten.affine_grid_generator', overload='default')>
- [ ] Implement <OpOverload(op='aten.aminmax', overload='default')>
- [ ] Implement <OpOverload(op='aten.binary_cross_entropy_with_logits', overload='default')>
- [ ] Implement <OpOverload(op='aten.bitwise_and', overload='Tensor')>
- [ ] Implement <OpOverload(op='aten.bucketize', overload='Tensor')>
- [ ] Implement <OpOverload(op='aten.conv_tbc', overload='default')>
- [ ] Implement <OpOverload(op='aten.fake_quantize_per_tensor_affine_cachemask', overload='default')>
- [ ] Implement <OpOverload(op='aten.fill', overload='Scalar')>
- [ ] Implement <OpOverload(op='aten.index_add', overload='default')>
- [ ] Implement <OpOverload(op='aten.index_copy', overload='default')>
- [ ] Implement <OpOverload(op='aten.index_fill', overload='int_Scalar')>
- [ ] Implement <OpOverload(op='aten.index_put', overload='default')>
- [ ] Implement <OpOverload(op='aten.masked_scatter', overload='default')>
- [ ] Implement <OpOverload(op='aten.masked_select', overload='default')>
- [ ] Implement <OpOverload(op='aten.prod', overload='dim_int')>
- [ ] Implement <OpOverload(op='aten.rsub', overload='Tensor')>
- [ ] Implement <OpOverload(op='aten.scatter', overload='src')>
- [ ] Implement <OpOverload(op='aten.scatter', overload='value')>
- [ ] Implement <OpOverload(op='aten.sort', overload='default')>
- [ ] Implement <OpOverload(op='aten.std', overload='correction')>
- [ ] Implement <OpOverload(op='aten.std_mean', overload='correction')>
- [ ] Implement <OpOverload(op='aten.sym_size', overload='int')>
- [ ] Implement <OpOverload(op='aten.take', overload='default')>
- Implement <OpOverload(op='aten._adaptive_avg_pool2d', overload='default')>
- Implement <OpOverload(op='aten._cdist_forward', overload='default')>
- Implement <OpOverload(op='aten._convolution', overload='default')>
- Implement <OpOverload(op='aten._fake_quantize_per_tensor_affine_cachemask_tensor_qparams', overload='default')>
- Implement <OpOverload(op='aten.grid_sampler_3d', overload='default')>
- Implement <OpOverload(op='aten.hann_window', overload='default')>
- Implement <OpOverload(op='aten.im2col', overload='default')>
- Implement <OpOverload(op='aten.repeat_interleave', overload='Tensor')>
- Implement <OpOverload(op='torchvision.nms', overload='default')>
- Implement <OpOverload(op='torchvision.roi_align', overload='default')>
- Implement <OpOverload(op='torchvision.roi_pool', overload='default')>
- [ ] Implement <OpOverload(op='aten.nan_to_num', overload='default')>
- [ ] Implement <OpOverload(op='aten.nll_loss2d_forward', overload='default')>
- [ ] Implement <OpOverload(op='aten.nll_loss_forward', overload='default')>
- [ ] Implement <OpOverload(op='aten.norm', overload='ScalarOpt_dim_dtype')>
- [ ] Implement <OpOverload(op='aten.pixel_unshuffle', overload='default')>

Add operator registration

- [ ] aten::empty
- [ ] aten::fill
- [ ] aten::getitem
- [ ] aten::normal
- [ ] aten::rsub
- [ ] aten::scatter_reduce
- [ ] aten::select
- [ ] aten::slice
- [ ] aten::softmax
- [ ] aten::subtract
- [ ] aten::transpose
- [ ] aten::unbind

### CODE REVIEW

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 1]
```diff
@@ -3571,7 +3571,7 @@ def aten_fmin(self: TensorType, other: TensorType) -> TensorType:
     raise NotImplementedError()
 
 
-@torch_op("aten::fmod")
+@torch_op(("aten::fmod", "aten::fmod.Tensor", "aten::fmod.Scalar"))
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:3574]: If an op does not have the `default` overload defined in https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/native_functions.yaml#L9722. We can remove the default version

```suggestion
@torch_op(("aten::fmod.Tensor", "aten::fmod.Scalar"))
```

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 2]
```diff
@@ -8531,10 +8531,11 @@ def aten_unsafe_chunk(self: TensorType, chunks: int, dim: int = 0) -> TensorType
     raise NotImplementedError()
 
 
-def aten_unsafe_split(self: TensorType, split_size: INT64, dim: int = 0) -> TensorType:
+@torch_op(("aten::unsafe_split", "aten::unsafe_split.Tensor"))
+def aten_unsafe_split(self: TTensor, split_size: INT64, dim: int = 0) -> TTensor:
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:8535]: Sequence[TTensor]

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 3]
```diff
@@ -7285,10 +7299,15 @@ def aten_rrelu(
...
+@torch_op(("aten::__rshift__.Tensor", "aten::__rshift__.Scalar"))
 def aten_rshift(self: TensorType, other: TensorType) -> TensorType:
     """__rshift__.Tensor(Tensor self, Tensor other) -> Tensor"""
 
-    raise NotImplementedError()
+    other = op.Cast(other, to=FLOAT.dtype)
+    two_pow = op.Pow(2.0, other)
+    two_pow = op.CastLike(two_pow, self)
+    rshift = op.Div(self, two_pow)
+    return rshift
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:7310]: This is bitwise shift?

File: onnxscript/function_libs/torch_lib/ops/linalg.py [hunk 4]
```diff
@@ -44,10 +44,28 @@ def aten_linalg_cond(self: TensorType, p: Optional[float] = None) -> TensorType:
     raise NotImplementedError()
 
 
-def aten_linalg_cross(self: TensorType, other: TensorType, dim: int = -1) -> TensorType:
-    """linalg_cross(Tensor self, Tensor other, *, int dim=-1) -> Tensor"""
+@torch_op("aten::linalg_cross")
+def aten_linalg_cross(self: TTensor, other: TTensor, dim: int = -1) -> TTensor:
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/linalg.py:48]: We can just register it under https://github.com/microsoft/onnxscript/blob/bcbf9f70afa6526478a57fe405ad464c4adfb631/onnxscript/function_libs/torch_lib/ops/core.py#L2240

File: onnxscript/function_libs/torch_lib/ops/linalg.py [hunk 5]
```diff
@@ -44,10 +44,28 @@ def aten_linalg_cond(self: TensorType, p: Optional[float] = None) -> TensorType:
...
 
-    raise NotImplementedError()
+    zero = op.Constant(value_ints=[0])
+    one = op.Constant(value_ints=[1])
+    two = op.Constant(value_ints=[2])
+    three = op.Constant(value_ints=[3])
+    axes = op.Expand(dim, op.Constant(value_ints=[1]))
```
Comment (gramalingam) [inline onnxscript/function_libs/torch_lib/ops/linalg.py:54]: This is converting a scalar to a 1D ... either Unsqueeze or Reshape is better than Expand. (It is more specific, and likely better understand by downstream analyses and optimizations.)

File: onnxscript/function_libs/torch_lib/ops/linalg.py [hunk 6]
```diff
@@ -44,10 +44,28 @@ def aten_linalg_cond(self: TensorType, p: Optional[float] = None) -> TensorType:
...
+
+    # Reference https://en.wikipedia.org/w/index.php?title=Cross_product&oldid=1143125073
+    a1 = op.Slice(self, zero, one, axes)
+    a2 = op.Slice(self, one, two, axes)
+    a3 = op.Slice(self, two, three, axes)
```
Comment (gramalingam) [inline onnxscript/function_libs/torch_lib/ops/linalg.py:59]: Is that axis expected to have a dimension exactly 3, or can it have more? If it is exactly 3, can we rewrite the above three as
```py
   a1, a2, a3 = op.Split(self, axis=dim, num_outputs=3)
```

File: onnxscript/function_libs/torch_lib/ops/linalg.py [hunk 7]
```diff
@@ -44,10 +44,28 @@ def aten_linalg_cond(self: TensorType, p: Optional[float] = None) -> TensorType:
...
+    b1 = op.Slice(other, zero, one, axes)
+    b2 = op.Slice(other, one, two, axes)
+    b3 = op.Slice(other, two, three, axes)
```
Comment (gramalingam) [inline onnxscript/function_libs/torch_lib/ops/linalg.py:62]: Same question as above

File: onnxscript/function_libs/torch_lib/ops/linalg.py (hunk 6 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/ops/linalg.py:59]: I think so. @shubhambhokare1 feel free to update my original implementation. Thanks!

File: onnxscript/function_libs/torch_lib/ops/linalg.py (hunk 4 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/ops/linalg.py:48]: I don’t think so. They are the exact implementation. One is alias of another

Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/ops/linalg.py:48]: Unless they are implemented differently?

**Review by justinchuby — APPROVED** [review]

File: tests/function_libs/torch_lib/ops_test_data.py [hunk 8]
```diff
@@ -2393,6 +2393,7 @@ def _where_input_wrangler(
     "imag",
     "isfinite",
     "le",
+    "lerp",
```
Comment (justinchuby) [inline tests/function_libs/torch_lib/ops_test_data.py:2396]: These are prims ops? Doesn’t look like this change is needed?

File: tests/function_libs/torch_lib/ops_test_data.py [hunk 9]
```diff
@@ -2402,6 +2403,7 @@ def _where_input_wrangler(
     "maximum",
     "minimum",
     "mul",
+    "mv",
```
Comment (justinchuby) [inline tests/function_libs/torch_lib/ops_test_data.py:2406]: Ditto

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 10]
```diff
@@ -4666,11 +4666,11 @@ def aten_le_bool(self: BOOL, other: BOOL) -> BOOL:
 def aten_lerp(self: TReal, end: TReal, weight: TReal) -> TReal:
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:4666]: If cast is needed, weight can take a different type than self

Comment (shubhambhokare1) [reply onnxscript/function_libs/torch_lib/ops/core.py]: Fixed, I was doing the casting incorrectly.

**Comment by justinchuby** [comment]
Good to merge. Thanks!
